File: ports/python/run_tests.py
# run_tests.py
# Driver for test_python.bat: reads ports/challenges.txt (one challenge per line),
# solves each challenge, and prints detailed per-challenge info to stdout.
# The batch file redirects this output to results.txt.

import os
import sys

# Make sure `challenge` is importable from this directory (challenge.py lives here).
HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)

import challenge  # noqa: E402


def parse_ts(challenge_str):
    """Parse the 8 leading hex chars of the challenge as a big-endian timestamp."""
    ts = 0
    for c in challenge_str[:8]:
        if '0' <= c <= '9':
            v = ord(c) - ord('0')
        elif 'a' <= c <= 'f':
            v = ord(c) - ord('a') + 10
        elif 'A' <= c <= 'F':
            v = ord(c) - ord('A') + 10
        else:
            v = 0
        ts = (ts << 4) | v
    return ts


def main():
    challenges_path = os.path.join(os.path.dirname(HERE), 'challenges.txt')

    with open(challenges_path, 'r', encoding='utf-8') as f:
        lines = [ln.strip() for ln in f if ln.strip()]

    for i, challenge_str in enumerate(lines, 1):
        ts = parse_ts(challenge_str)
        ts_hex = challenge_str[:8]
        key = challenge.decode_key(challenge_str)
        dc = challenge.day_counter(ts)
        pp = ((dc - 1) % 99) + 1  # the 2-digit prefix used in the response
        m1, m2 = challenge.mac_compute(ts, key) if key else (b'', b'')
        response = challenge.solve(challenge_str)

        print('=' * 72)
        print(f'Challenge #{i}')
        print('=' * 72)
        print(f'  challenge : {challenge_str}')
        print(f'  timestamp : 0x{ts_hex}  (= {ts} decimal)')
        print(f'  day-counter: {dc}')
        print(f'  response prefix (pp): {pp:02d}')
        print(f'  key (M1)  : {key!r}  (len={len(key)})')
        if m1:
            print(f'  M1 (MD5)  : {challenge._bytes_to_hex(m1)}')
            print(f'  M2 (MD5)  : {challenge._bytes_to_hex(m2)}')
        print(f'  response  : {response}')
        print()


if __name__ == '__main__':
    main()

File: ports/lua5.1/run_tests.lua
-- File: run_tests.lua
-- Driver for test_lua.bat: reads ../challenges.txt (one challenge per line),
-- solves each challenge, and prints detailed per-challenge info to stdout.
-- The batch file redirects this output to results.txt.

package.path = 'lua_modules/?.lua;' .. package.path

local script_dir = arg[0] and arg[0]:match('^(.*)[/\\]') or '.'
local challenge_path = script_dir .. '/challenge.lua'

-- Load the challenge module via dofile so we can call its exported `solve`,
-- and expose internal helpers we need (decode_key, day_counter, mac_compute).
-- We re-require the internals by loading with a table that exposes locals is not
-- possible, so we make challenge.lua expose the extra helpers for the test driver.
-- To keep challenge.lua's public API clean, this driver depends on the extra
-- debug exports below. See note at bottom.

local challenge = dofile(challenge_path)

-- The challenge module only exports `solve`. For detailed per-challenge output
-- we re-implement parse_ts and day_counter here locally (they are trivial and
-- mirror challenge.hpp), and derive the key indirectly is NOT possible without
-- the internal decode_key. Therefore we add a small extension: challenge.lua
-- exports `_debug = { decode_key, day_counter, mac_compute, bytes_to_hex }`.

local dbg = challenge._debug
if not dbg then
    io.stderr:write('ERROR: challenge.lua does not export _debug helpers.\n')
    os.exit(1)
end

local function parse_ts(challenge_str)
    local ts = 0
    for i = 1, 8 do
        local c = challenge_str:sub(i, i)
        local v = tonumber(c, 16) or 0
        ts = ts * 16 + v
    end
    return ts
end

local challenges_file = script_dir .. '/../challenges.txt'
local f = io.open(challenges_file, 'r')
if not f then
    io.stderr:write('ERROR: cannot open ' .. challenges_file .. '\n')
    os.exit(1)
end
local lines = {}
for line in f:lines() do
    line = line:gsub('^%s+', ''):gsub('%s+$', '')
    if line ~= '' then lines[#lines + 1] = line end
end
f:close()

for idx, challenge_str in ipairs(lines) do
    local ts = parse_ts(challenge_str)
    local ts_hex = challenge_str:sub(1, 8)
    local key = dbg.decode_key(challenge_str)
    local dc = dbg.day_counter(ts)
    local pp = ((dc - 1) % 99) + 1
    local response = challenge.solve(challenge_str)

    print(string.rep('=', 72))
    print(string.format('Challenge #%d', idx))
    print(string.rep('=', 72))
    print('  challenge : ' .. challenge_str)
    print(string.format('  timestamp : 0x%s  (= %d decimal)', ts_hex, ts))
    print(string.format('  day-counter: %d', dc))
    print(string.format('  response prefix (pp): %02d', pp))
    if key then
        print(string.format('  key (M1)  : %q  (len=%d)', key, #key))
        local m1, m2 = dbg.mac_compute(ts, key)
        if m1 then
            print('  M1 (MD5)  : ' .. dbg.bytes_to_hex(m1))
            print('  M2 (MD5)  : ' .. dbg.bytes_to_hex(m2))
        end
    else
        print('  key (M1)  : <failed to decode>')
    end
    print('  response  : ' .. tostring(response))
    print()
end
